### PR TITLE
buildGoModule: make the vendor fetcher error if it is empty

### PR DESCRIPTION
**Pull request by zowoq**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @NixOS/golang 

See discussion in https://github.com/NixOS/nixpkgs/pull/126682 and https://github.com/NixOS/nixpkgs/pull/211962

Draft as still need to check this is fine with proxyVendor, etc.

### CODE REVIEW

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by zowoq** [comment]
@ofborg build evmdis lnch phylactery statik vimPlugins.vim-hexokinase

**Comment by qbit** [comment]
Result of `nixpkgs-review pr 212024` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>evmdis</li>
    <li>lnch</li>
    <li>phylactery</li>
    <li>statik</li>
    <li>vimPlugins.vim-hexokinase</li>
  </ul>
</details>

**Comment by qbit** [comment]
Result of `nixpkgs-review pr 212024` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>evmdis</li>
    <li>lnch</li>
    <li>phylactery</li>
    <li>statik</li>
    <li>vimPlugins.vim-hexokinase</li>
  </ul>
</details>

File: pkgs/build-support/go/module.nix [hunk 1]
```diff
@@ -145,6 +145,11 @@ let
       cp -r --reflink=auto vendor $out
     ''}
 
+      if ! [ "$(ls -A $out)" ]; then
+        echo "vendor folder is not needed, please set 'vendorHash = null;' or 'vendorSha256 = null;' in your expression"
```
Comment (asymmetric) [inline pkgs/build-support/go/module.nix:149]:
```suggestion
        echo "vendor folder is empty, please set 'vendorHash = null;' or 'vendorSha256 = null;' in your expression"
```

**Review by asymmetric — APPROVED** [review]
Just a nit.

File: pkgs/build-support/go/module.nix (hunk 1 above)
Comment (asymmetric) [reply pkgs/build-support/go/module.nix:149]: But I'm actually unsure if it's an improvement, I like yours as well.

Comment (zowoq) [reply pkgs/build-support/go/module.nix:149]: I'll switch to empty as it is more consistent with the other vendor error messages.

**Pull request merged**